### PR TITLE
[TACHYON-590] Utility functions for the wildcard support of Tachyon shell commands

### PR DESCRIPTION
**Pull request by tianyin**
There are two public methods, 
`getTachyonURIs` and `getFiles`
; the former is for Tachyon files and the latter is for files in the local fs.

Basically, the method takes a input path (could contain wildcards) and return the list of files that matches the input path (if the input path is an absolute path, it returns the path)

### CODE REVIEW

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java [hunk 1]
```diff
@@ -74,4 +80,163 @@ public static String validatePath(String path, TachyonConf tachyonConf) throws I
...
+   * Else if the path contains wildcards, the returned list contains all the matched URIs
+   * It supports any number of wildcards in inputURI
+   * @param tachyonClient The client that fetches the paths
+   * @param inputURI the input URI (could contain wildcards)
+   * @return A list of TachyonURI that matches the inputURI
+   * @throws IOException
+   */
+  public static List<TachyonURI> getTachyonURIs(TachyonFS tachyonClient, TachyonURI inputURI)
+      throws IOException {    
+    if (inputURI.getPath().contains(TachyonURI.WILDCARD) == false) {
```
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: A better style would be `if (!inputURI.getPath().contains(TachyonURI.WILDCARD))`

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java [hunk 2]
```diff
@@ -74,4 +80,163 @@ public static String validatePath(String path, TachyonConf tachyonConf) throws I
...
+   * @throws IOException
+   */
+  private static List<TachyonURI> getTachyonURIs(TachyonFS tachyonClient,
+      TachyonURI inputURI, TachyonURI parentDir) throws IOException {
+    List<TachyonURI> res = new LinkedList<TachyonURI>(); 
+    List<ClientFileInfo> files = tachyonClient.listStatus(parentDir);
+    for (ClientFileInfo file : files) {
+      TachyonURI fileURI = new TachyonURI(inputURI.getScheme(), 
+                                          inputURI.getAuthority(), 
+                                          file.getPath());
```
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: Please fix the formatting here and in other parts of the PR.

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java [hunk 3]
```diff
@@ -74,4 +80,163 @@ public static String validatePath(String path, TachyonConf tachyonConf) throws I
...
+      if (match(fileURI, inputURI)) { //if it matches
+        res.add(fileURI);
+      } else {
+        if (file.isFolder) { //if it is a folder, we do it recursively
+          TachyonURI dirURI = new TachyonURI(inputURI.getScheme(), 
+                                             inputURI.getAuthority(), 
+                                             file.getPath());
+          String prefix = inputURI.getLeadingPath(dirURI.getDepth());
+          if (prefix != null && match(dirURI, new TachyonURI(prefix)) == true) {
```
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: You shouldn't need the `== true`

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java [hunk 4]
```diff
@@ -74,4 +80,163 @@ public static String validatePath(String path, TachyonConf tachyonConf) throws I
...
+   * Basically, it recursively iterates through the directory from the parent directory 
+   * (e.g., for input "/a/b/*", it will start from "/a/b") until it finds all the matches;
+   * It does not go into a directory if the prefix mismatches
+   * (e.g., for input "/a/b/*", it won't go inside directory "/a/c")
+   * If a file path matches 
+   * @param tachyonClient
+   * @param inputURI
+   * @param parentDir
+   * @return
+   * @throws IOException
```
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: Please complete the incomplete java docs in this PR.

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java [hunk 5]
```diff
@@ -74,4 +80,163 @@ public static String validatePath(String path, TachyonConf tachyonConf) throws I
...
+   * The utility function used to implement getFiles
+   * It follows the same algorithm as getTachyonURIs
+   * @param inputPath
+   * @param parent
+   * @return
+   */
+  protected static List<File> getFiles(String inputPath, String parent) {
+    List<File> res = new LinkedList<File>();
+    File pFile = new File(parent);
+    if (pFile.exists() == false || pFile.isDirectory() == false) {
```
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: This can be simplified (similar to the other conditions).

**Comment by calvinjia** [comment]
@tianyin Thanks for the patch, I left a few comments.

**Comment by tianyin** [comment]
thanks @calvinjia ! Just for my curiosity, the condition simplification (eliminating true/false) is a style thing or other reasons?

**Comment by calvinjia** [comment]
@tianyin Its mostly a style thing, doing `false == false` to get `true` is the same as `!false` which is more concise and easier to read I think.

**Comment by tianyin** [comment]
@calvinjia got it!!

**Comment by calvinjia** [comment]
+1 LGTM

File: shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java [hunk 6]
```diff
@@ -42,4 +69,175 @@ public void getFilePathTest() throws IOException {
...
+    List<File> fl5 = TFsShellUtils.getFiles(rootDir + "/*/foo*");
+    Collections.sort(fl5, createFilePathComparator());
+    Assert.assertEquals(fl5.size(), 3);
+    assertFilePath(fl5.get(0), rootDir + "/bar/foobar3");
+    assertFilePath(fl5.get(1), rootDir + "/foo/foobar1");
+    assertFilePath(fl5.get(2), rootDir + "/foo/foobar2");
+  }
+  
+  @Test
+  public void matchTest() {
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java]: What if `match("/a/b/c", "*")`?

File: shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java [hunk 7]
```diff
@@ -42,4 +69,175 @@ public void getFilePathTest() throws IOException {
...
+     * Generate such local structure
+     *  /testWildCards
+     *  ├── foo
+     *  |    ├── foobar1
+     *  |    └── foobar2
+     *  ├── bar
+     *  |    └── foobar3
+     *  └── foobar4
+     */
+    FileUtils.deleteDirectory(new File(mLocalTachyonCluster.getTachyonHome() + "/testWildCards"));
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java]: `TemporaryFolder` can be used to create a temporary folder here, the folder will be automatically deleted from local filesystem after the unit test, example is here: https://github.com/amplab/tachyon/blob/master/servers/src/test/java/tachyon/worker/block/evictor/EvictorTestBase.java#L47

File: shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java [hunk 8]
```diff
@@ -42,4 +69,175 @@ public void getFilePathTest() throws IOException {
...
+        TachyonFSTestUtils.createByteFile(mTfs, "/testWildCards/bar/foobar3", 
+            WriteType.MUST_CACHE, 30);
+    fileIds[3] =
+        TachyonFSTestUtils.createByteFile(mTfs, "/testWildCards/foobar4", 
+            WriteType.MUST_CACHE, 40);
+    return fileIds;
+  }
+  
+  @Test
+  public void getTachyonURIsTest() throws IOException {
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java]: The logic of this test is the same as `getFilesTest()`, what about having utility methods like 
```
enum FsType {
    TFS,
    LOCAL
}; 

List<String> getPaths(String path, FsType fsType) {
     List<String> ret;
     if (fsType == FsType.TFS) {
         List<TachyonURI> tPaths = TFsShellUtils.getTachyonURIs(mTfs, new TachyonURI(path));
         ret = new ArrayList<String>(tPaths.size());
         for (TachyonURI tPath : tPaths) {
             ret.add(tPath.getPath());
         }
    } else {
      .....
   }
}

void sortPaths(List<String> paths) {
    ....
}

void resetFsHierarchy(FsType fsType) {
   ....
}
```

and then we can just have one `getPathTest` like 
```
void getPathTest() {
    for (FsType fsType : FsType.values()) {
       resetFsHierarchy(fsType);
       List<String> paths = getPaths(path, fsType);
       sortPaths(paths);
       .....
   }
}
```

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java [hunk 9]
```diff
@@ -74,4 +80,160 @@ public static String validatePath(String path, TachyonConf tachyonConf) throws I
...
+  }
+  
+  /**
+   * The utility function used to implement getFiles
+   * It follows the same algorithm as getTachyonURIs
+   * @param inputPath The input file path (could contain wildcards)
+   * @param parent The directory in which we are searching matched files 
+   * @return A list of files that matches the input path in the parent directory
+   */
+  protected static List<File> getFiles(String inputPath, String parent) {
```
Comment (uronce-cc) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: `private` to be the same with `getTachyonURIs`?

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java [hunk 10]
```diff
@@ -74,4 +80,160 @@ public static String validatePath(String path, TachyonConf tachyonConf) throws I
...
+    return escape(text).replace("\\*", ".*");
+  }
+  
+  /**
+   * Return whether or not fileURI matches the patternURI
+   * @param fileURI The TachyonURI of a particular file
+   * @param patternURI The URI that can contain wildcards
+   * @return true if matches; false if not
+   */
+  protected static boolean match(TachyonURI fileURI, TachyonURI patternURI) {
```
Comment (uronce-cc) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: `private`, seems no need to be `protected`.

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java [hunk 11]
```diff
@@ -74,4 +80,160 @@ public static String validatePath(String path, TachyonConf tachyonConf) throws I
...
+    return escape(fileURI.getPath()).matches(replaceWildcards(patternURI.getPath()));
+  }
+  
+  /**
+   * Return whether or not filePath matches patternPath
+   * @param filePath Path of a given file
+   * @param patternPath Path that can contain wildcards
+   * @return true if matches; false if not
+   */
+  protected static boolean match(String filePath, String patternPath) {
```
Comment (uronce-cc) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: `private` ? seems no need to be `protected`.

**Comment by uronce-cc** [comment]
@tianyin lgtm overall, just left few minor comments and a comment to remove duplicated test codes.

**Comment by tianyin** [comment]
@uronce-cc got it! I will revise asap! Thanks a lot, Chang!

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java (hunk 11 above)
Comment (tianyin) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: @uronce-cc if I change this one to private, I cannot call it in the TfsShellUtilsTest :-(

File: shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java (hunk 6 above)
Comment (tianyin) [inline shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java]: @uronce-cc you mean add these two test cases here? Or?

Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellUtilsTest.java]: yeah.

File: shell/src/main/java/tachyon/shell/TFsShellUtils.java (hunk 11 above)
Comment (uronce-cc) [inline shell/src/main/java/tachyon/shell/TFsShellUtils.java]: Got it.

**Comment by tianyin** [comment]
@uronce-cc I addressed your review comments. Take a look :-)

**Comment by uronce-cc** [comment]
+1 LGTM!

**Pull request merged**